File: 443.String-Compression.go
// https://leetcode.com/problems/string-compression/description/
//
// algorithms
// Easy (35.8%)
// Total Accepted:    29.7k
// Total Submissions: 82.9k
// beats 100.0% of golang submissions

package leetcode

import "strconv"

func compress(chars []byte) int {
	ch, num, idx := chars[0], 0, 0

	for _, char := range chars {
		if char == ch {
			num++
		} else {
			chars[idx] = ch
			idx++
			if num > 1 {
				for _, n := range strconv.Itoa(num) {
					chars[idx] = byte(n)
					idx++
				}
			}
			ch = char
			num = 1
		}
	}

	chars[idx] = ch
	idx++
	if num > 1 {
		for _, n := range strconv.Itoa(num) {
			chars[idx] = byte(n)
			idx++
		}
	}

	return len(chars[:idx])
}

File: 443.String-Compression.java
// https://leetcode.com/problems/string-compression/
//
// algorithms
// Easy (38.45%)
// Total Accepted:    67,153
// Total Submissions: 174,636
// beats 100.0% of java submissions


class Solution {
    public int compress(char[] chars) {
        int length = chars.length;
        int i = 0, j = -1, idx = 0;
        
        while (i < length) {
            j = i + 1;
            while (j < length && chars[j] == chars[i]) {
                j += 1;
            }
            
            int repeatNum = j - i;
            if (repeatNum == 1) {
                chars[idx++] = chars[i];
            } else {
                chars[idx++] = chars[i];
                char[] tmp = String.valueOf(repeatNum).toCharArray();
                for (int k = 0; k < tmp.length; k++) {
                    chars[idx++] = tmp[k];
                }
            }
            
            i = j;
        }
        
        return idx;
    }
}